spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java

```diff
@@ -0,0 +1,197 @@
+  private final FileIO fileIO;
+  private final int partitionDiscoveryParallelism;
+  private final String dataLocation;
+
+  private String allDataFilesTable = null;
+  private Long olderThanTimestamp = null;
+  private boolean dryRun = false;
+
+  RemoveOrphanFilesAction(SparkSession spark, Table table) {
+    this.spark = spark;
+    this.sparkContext = new JavaSparkContext(spark.sparkContext());
+    this.hadoopConf = new SerializableConfiguration(spark.sessionState().newHadoopConf());
+    this.fileIO = table.io();
+    this.partitionDiscoveryParallelism = spark.sessionState().conf().parallelPartitionDiscoveryParallelism();
+    this.dataLocation = table.locationProvider().dataLocation();
+  }
+
+  public RemoveOrphanFilesAction allDataFilesTable(String newAllDataFilesTable) {
+    this.allDataFilesTable = newAllDataFilesTable;
+    return this;
+  }
+
+  public RemoveOrphanFilesAction olderThan(long newOlderThanTimestamp) {
+    this.olderThanTimestamp = newOlderThanTimestamp;
+    return this;
+  }
+
+  public RemoveOrphanFilesAction dryRun(boolean newDryRun) {
+    this.dryRun = newDryRun;
+    return this;
+  }
+
+  @Override
+  public RemoveOrphanFilesActionResult execute() {
+    Preconditions.checkArgument(allDataFilesTable != null, "allDataFilesTable must be set");
+    Preconditions.checkArgument(olderThanTimestamp != null, "olderThanTimestamp should be set");
+
+    Dataset<Row> validDataFileDF = buildValidDataFileDF();
+    Dataset<Row> actualDataFileDF = buildActualDataFileDF();
+
+    Column joinCond = validDataFileDF.col("file_path").equalTo(actualDataFileDF.col("file_path"));
+    List<String> orphanDataFiles = actualDataFileDF.join(validDataFileDF, joinCond, "leftanti")
+        .as(Encoders.STRING())
+        .collectAsList();
+
+    if (!dryRun) {
+      Tasks.foreach(orphanDataFiles)
+          .noRetry()
+          .suppressFailureWhenFinished()
+          .onFailure((file, exc) -> LOG.warn("Failed to delete data file: {}", file, exc))
+          .run(fileIO::deleteFile);
+    }
+
+    return new RemoveOrphanFilesActionResult(orphanDataFiles);
+  }
+
+  private Dataset<Row> buildValidDataFileDF() {
+    return spark.read().format("iceberg")
+        .load(allDataFilesTable)
+        .select("file_path")
+        .distinct();
+  }
+
+  private Dataset<Row> buildActualDataFileDF() {
+    List<String> topLevelDirs = Lists.newArrayList();
+    List<String> matchingTopLevelFiles = Lists.newArrayList();
+
+    try {
+      Path dataPath = new Path(dataLocation);
+      FileSystem fs = dataPath.getFileSystem(hadoopConf.value());
+
+      for (FileStatus file : fs.listStatus(dataPath, HiddenPathFilter.get())) {
+        // TODO: handle custom metadata folders
+        // we need to ignore the metadata folder when data is written to the root table location
+        if (file.isDirectory() && !"metadata".equals(file.getPath().getName())) {
+          topLevelDirs.add(file.getPath().toString());
```

Comment:
I would probably opt not to do this check. Instead, there is little cost to doing the anti-join with all known data and metadata file paths. I'd probably do that instead of worrying about metadata location.

Reply:
Done.